### DISCUSSION TITLE
Point time calculation

### DISCUSSION THREAD

**Discussion by vovo-4K** [ouster-lidar/ouster-ros#206] [Q&A] [answered]
There is the **t** field in the Point structure:

```c++
struct EIGEN_ALIGN16 Point {
    PCL_ADD_POINT4D;
    float intensity;
    uint32_t t;
    uint16_t reflectivity;
    uint16_t ring;
    uint16_t ambient;
    uint32_t range;
    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
};
```

https://github.com/ouster-lidar/ouster-ros/blob/19071eb14e0c8fecf0d94d1b2399aea5f44d549c/ouster-ros/include/ouster_ros/os_point.h#L23C16-L23C16

How can I calculate absolute point time using message header time and the **t** field?

**Comment by Samahu** [comment]
Simple, take the value of `header.stamp` from the published **sensor_msgs::msg::PointCloud2** message and add it to the value of `t` field from **ouster_ros::Point**.

**Reply by vovo-4K** [reply]
Thank you for your reply! The header.stamp is composed of two parts: an integer representing seconds and an unsigned integer representing nanoseconds. What is the `ouster_ros::Point::t` field dimension? Does it contain nanoseconds offset since header.stamp?

**Reply by Samahu** [reply, marked as answer]
`ouster_ros::Point::t` is represented in nanoseconds.

**Reply by liuXinGangChina** [reply]
nice answer,
by the way , do you mean the sbsolute time stamp of a point = lidar.header.stamp.sec +  lidar.header.stamp.nsec x 10^6 + t x 10^6
@Samahu

**Reply by Samahu** [reply]
I think the correct formula would be: `pt_absolute_ts_ns = lidar.header.stamp.sec * 10^9 + lidar.header.stamp.nsec + pt.t` then you can convert that amount back to seconds if you choose to

**Reply by liuXinGangChina** [reply]
thanks got it !
